core/src/main/scala/scheduler/K8sHttp.scala

```diff
@@ -0,0 +1,156 @@
+
+    val request = addCreds(dc, Request(
+      Method.DELETE,
+      k8s.endpoint / "apis" / "apps" / "v1beta2" / "namespaces" / "default" / "deployments" / name
+    ))
+
+    // TODO: Potential 404 or similar here if trying to delete something that doesn't exist
+    client.expect[Json](request).map(_ => ())
+  }
+
+  def launch(image: Image, dc: Datacenter, ns: NamespaceName, unit: UnitDef, version: Version, plan: Plan, hash: String): Task[String] = {
+    val stackName = StackName(unit.name, version, hash)
+    val name = buildName(stackName)
+    val env = plan.environment.bindings ++ List(
+      EnvironmentVariable("NELSON_STACKNAME",        stackName.toString),
+      EnvironmentVariable("NELSON_DATACENTER",       dc.name),
+      EnvironmentVariable("NELSON_ENV",              ns.root.asString),
+      EnvironmentVariable("NELSON_NAMESPACE",        ns.asString),
+      EnvironmentVariable("NELSON_DNS_ROOT",         dc.domain.name),
+      EnvironmentVariable("NELSON_PLAN",             plan.name),
+      EnvironmentVariable("NELSON_DOCKER_IMAGE",     image.toString),
+      EnvironmentVariable("NELSON_MEMORY_LIMIT",     plan.environment.memory.getOrElse(512D).toInt.toString),
+      EnvironmentVariable("NELSON_NODENAME",         s"$${node.unique.name}"),
+      EnvironmentVariable("NELSON_VAULT_POLICYNAME", NomadJson.getPolicyName(ns, name))
+    )
+    val newPlan = plan.copy(environment = plan.environment.copy(bindings = env))
+    val json = K8sJson.deployment(name, image, plan, unit.ports)
+
+    val request = addCreds(dc, Request(
+      Method.POST,
+      k8s.endpoint / "apis" / "apps" / "v1beta2" / "namespaces" / "default" / "deployments"
+    ))
+
+    for {
+      _        <- Task.delay(log.debug(s"Sending K8s the following payload: ${json.nospaces}"))
+      response <- client.expect[Json](request.withBody(json))
+    } yield response.nospaces
+  }
+
+  private def buildName(sn: StackName): String = sn.toString
+
+  private def addCreds(dc: Datacenter, req: Request): Request =
+    req.putHeaders(Authorization(Token(AuthScheme.Bearer, serviceAccountToken)))
+}
+
+object K8sJson {
+  def deployment(
+    name:     String,
+    // unitName: UnitName,          // ignored
+    image:    Image,
+    // dc:       Datacenter,        // ignored
+    // ns:       NamespaceName,     // ignored
+    plan:     Plan,
+    // schedule: Option[Schedule],  // ignored
+    ports:    Option[Ports]
+  ): Json = {
+    // TODO: Cron/Job not supported atm, only Service
+    // TODO: Fix ports
+    Json(
+      "apiVersion" := "apps/v1beta2",
+      "kind"       := "Deployment",
+      "metadata"   := Json(
+        "name"   := name,
+        "labels" := Json(
+          "app"    := name,
+          "nelson" := "true"
+        )
+      ),
+      "spec" := Json(
+        "replicas" := plan.environment.desiredInstances.getOrElse(1),
+        "selector" := Json(
+          "matchLabels" := Json("app" := name)
```

Comment:
Kubernetes has a notion of [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) for each pod (analogous to a unit of deployment, could be a service, a service + sidecars), this `matchLabels` determines which pods are controlled by this deployment. This becomes important since the pods deployed through a Kubernetes Deployment aren't special, there could be existing pods that match a label in which case the Deployment will consider those as part of it's control.

For example if there is an existing `foo` with label `name=foo`, and you create a Deployment for `foo` with `"matchLabels": { "name": "foo" }` for 2 replicas, it will notice there is already an existing one and only create one additional pod.

tl;dr `matchLabels` should match on labels that uniquely identify a particular deployment, so I think `name` (which is the stack name) makes sense here, do I need anything else?